.cursor/rules/data-encoding.mdc

```diff
@@ -0,0 +1,292 @@
+        }
+        // Handle raw CLI input
+        TriggerData::Raw(data) => Ok((0, data.clone(), Destination::CliOutput)),
+        _ => Err(anyhow::anyhow!("Unsupported trigger data type")),
+    }
+}
+```
+
+2. Raw Input Processing:
+```rust
+// Convert bytes to string
+let input = std::str::from_utf8(&req)
+    .map_err(|e| e.to_string())?;
+
+// Parse first character as hex digit
+let value = input
+    .chars()
+    .next()
+    .ok_or("Empty input")?
+    .to_digit(16)
+    .ok_or("Invalid hex digit")? as u64;
+```
+
+## Output Encoding
+
+1. Basic Output Encoding:
+```rust
+fn encode_trigger_output(trigger_id: u64, output: impl AsRef<[u8]>) -> Vec<u8> {
+    solidity::DataWithId {
+        triggerId: trigger_id,
+        data: output.as_ref().to_vec().into()
+    }.abi_encode()
+}
+```
+
+2. Destination-based Encoding:
+```rust
+let output = match dest {
+    // For Ethereum: Encode with trigger ID
+    Destination::Ethereum => Some(encode_trigger_output(trigger_id, &result)),
+    // For CLI: Return raw output
+    Destination::CliOutput => Some(result),
+};
+```
+
+## Common Patterns
+
+1. Event Data Decoding:
+```rust
+// Decode Solidity string from event data
+let decoded = alloy_sol_types::sol_data::String::abi_decode(&log.data, false)
+    .map_err(|e| format!("Failed to decode ABI string: {}", e))?;
+
+let message = decoded.to_string();
+```
+
+2. Bridge Event Handling:
+```rust
+// Handle different event types
+match trigger_data {
+    TriggerData::EthContractEvent(event) => {
+        // Decode Ethereum event
+        let event: solidity::NewTrigger = decode_event_log_data!(log)?;
+    }
+    TriggerData::CosmosContractEvent(event) => {
+        // Parse Cosmos event attributes
+        if event.ty.as_str() == NewDepositEvent::KEY {
+            for (key, value) in event.attributes {
+                match key.as_str() {
+                    "amount" => amount = Some(value),
+                    "sender" => sender = Some(value),
+                    "recipient" => recipient = Some(value),
+                }
+            }
+        }
+    }
+}
+```
+
+## Best Practices
+
+1. Input Validation:
+- Always validate decoded input before processing
+- Handle empty or malformed input gracefully
+- Provide clear error messages for decoding failures
+
+2. Error Handling:
+```rust
+// Use descriptive error messages
+.map_err(|e| format!("Failed to decode event log data: {}", e))?;
+
+// Handle specific error cases
+match decode_result {
+    Ok(data) => process_data(data),
+    Err(e) => handle_error(e),
+}
+```
+
+3. Type Safety:
+- Use generated Solidity types for contract interactions
+- Validate numeric conversions
+- Check array bounds and lengths
+- Verify address formats
+
+4. Output Formatting:
+```rust
+// For Ethereum submission
+let payload = TransactionPayload {
+    data: result.abi_encode(),
+    destination: validated_address,
+    value: amount,
+};
```

Comment:
This is specific to an example. Would just kill the Output Formatting section for now. See my comment above about explaining destinations and outputs better so we know the encoding required for CLI output and EVM output.

Reply:
Removed